File: screen-reader/android/src/main/java/com/capacitorjs/plugins/screenreader/ScreenReaderPlugin.kt
package com.capacitorjs.plugins.screenreader

import com.getcapacitor.JSObject
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginMethod
import com.getcapacitor.annotation.CapacitorPlugin

@CapacitorPlugin(name = "ScreenReader")
public class ScreenReaderPlugin : Plugin() {
    private lateinit var screenReader: ScreenReader

    override fun load() {
        screenReader = ScreenReader(context)
        screenReader.addStateChangeListener { enabled ->
            val ret = JSObject()
            ret.put("value", enabled)
            notifyListeners(EVENT_STATE_CHANGE, ret)
        }
    }

    override fun handleOnDestroy() {
        screenReader.removeAllListeners()
    }

    @PluginMethod
    public fun isEnabled(call: PluginCall) {
        val ret = JSObject()
        ret.put("value", screenReader.isEnabled)
        call.resolve(ret)
    }

    @PluginMethod
    public fun speak(call: PluginCall) {
        val value = call.getString("value")
        val language = call.getString("language") ?: "en"
        screenReader.speak(value, language)
        call.resolve()
    }

    public companion object {
        public const val EVENT_STATE_CHANGE: String = "stateChange"
    }
}

File: screen-reader/ios/Sources/ScreenReaderPlugin/ScreenReaderPlugin.swift
import Foundation
import UIKit
import Capacitor

@objc(ScreenReaderPlugin)
public class ScreenReaderPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "ScreenReaderPlugin"
    public let jsName = "ScreenReader"
    public let pluginMethods: [CAPPluginMethod] = [
        .promise("speak", ScreenReaderPlugin.speak),
        .async("isEnabled", ScreenReaderPlugin.isEnabled)
    ]
    static let stateChangeEvent = "stateChange"

    override public func load() {
        NotificationCenter.default.addObserver(self,
                                               selector: #selector(self.onVoiceOverStateChanged(notification:)),
                                               name: UIAccessibility.voiceOverStatusDidChangeNotification,
                                               object: nil)
    }

    deinit {
        NotificationCenter.default.removeObserver(self)
    }

    /// UIAccessibility is UIKit state: the method runs on the main actor, not the bridge queue.
    @MainActor
    func isEnabled(_ call: CAPPluginCall) async -> JSObject {
        return [
            "value": UIAccessibility.isVoiceOverRunning
        ]
    }

    /// Stays synchronous so that announcements keep the order of the calls, which async methods do not. The
    /// announcement is made on the main queue a moment later, as before, and VoiceOver is checked there: it used to be
    /// read on the bridge queue.
    func speak(_ call: CAPPluginCall) throws {
        guard let value = call.getString("value") else {
            throw CAPPluginError("No value provided")
        }

        DispatchQueue.main.asyncAfter(deadline: .now() + 0.1) {
            if UIAccessibility.isVoiceOverRunning {
                UIAccessibility.post(notification: .announcement, argument: value)
            }
        }

        call.resolve()
    }

    @objc private func onVoiceOverStateChanged(notification: NSNotification) {
        notifyListeners(ScreenReaderPlugin.stateChangeEvent, data: [
            "value": UIAccessibility.isVoiceOverRunning
        ])
    }
}
